src/EventStore.Core/TransactionLog/Chunks/TFChunkWriter.cs

```diff
@@ -40,24 +43,50 @@ public class TFChunkWriter : ITransactionFileWriter {
 			// DO NOTHING
 		}
 
+		public void Begin() {
+			if (_inTransaction)
+				throw new Exception("Attempt to begin a new transaction while there is an ongoing transaction.");
+
+			_inTransaction = true;
+		}
+
 		public bool Write(ILogRecord record, out long newPos) {
 			var result = _currentChunk.TryAppend(record);
 			if (result.Success)
-				_writerCheckpoint.Write(result.NewPosition + _currentChunk.ChunkHeader.ChunkStartPosition);
+				_nextRecordPosition = result.NewPosition + _currentChunk.ChunkHeader.ChunkStartPosition;
 			else
-				CompleteChunk(); // complete updates checkpoint internally
-			newPos = _writerCheckpoint.ReadNonFlushed();
+				CompleteChunk();
+
+			if (!_inTransaction)
```

Comment:
it took me a while to put my finger on what was making me uncertain about this, but:   i think if we want to give the user of TFChunkWriter a way to use it without having to Open and Commit transactions, then we should make a separate write method for that.

Or more likely have a separate method for writing to a transaction and then have the regular Write method call OpenTransaction, WriteToTransaction and CommitTransaction.

The risk with the way that it is here currently is that if something has somehow left a transaction open, then the EpochManager and PartitionManager will just call Write and Flush, expecting the data to then be safely on disk, but in fact it will just be added to the open transaction and not flushed (and no problem detected)

Reply:
great point, fixed :+1: